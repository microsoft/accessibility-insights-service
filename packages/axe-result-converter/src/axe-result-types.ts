// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import axe from 'axe-core';

export declare type SelectorType = 'xpath' | 'css';

export interface AxeCoreResults extends axe.AxeResults {
    urls: string[];
    passes: AxeResult[];
    violations: AxeResult[];
    incomplete: AxeResult[];
    inapplicable: AxeResult[];
}

export interface AxeResult extends axe.Result {
    urls: string[];
    nodes: AxeNodeResult[];
}

export interface AxeNodeResult extends axe.NodeResult {
    selectors: Selector[];
    fingerprint: string;
}

export interface Selector {
    selector: string;
    type: SelectorType;
}
