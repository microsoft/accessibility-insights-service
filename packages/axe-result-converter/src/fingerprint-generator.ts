// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable, inject } from 'inversify';
import { HashGenerator } from 'common';

/* eslint-disable @typescript-eslint/no-explicit-any */

export interface FingerprintParameters {
    rule: string;
    snippet: string;
    cssSelector: string;
    xpathSelector?: string;
}

@injectable()
export class FingerprintGenerator {
    constructor(@inject(HashGenerator) private readonly hashGenerator: HashGenerator) {}

    public getFingerprint(fingerprintParameters: FingerprintParameters): string {
        const valuesToHash = [fingerprintParameters.rule, fingerprintParameters.snippet, fingerprintParameters.cssSelector];

        if (fingerprintParameters.xpathSelector) {
            valuesToHash.push(fingerprintParameters.xpathSelector);
        }

        return this.hashGenerator.generateBase64Hash(...valuesToHash);
    }
}
