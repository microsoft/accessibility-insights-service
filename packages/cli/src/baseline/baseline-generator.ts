// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { AxeResult, AxeResultsList } from 'axe-result-converter';
import { injectable } from 'inversify';
import { identity, sortBy } from 'lodash';
import { BaselineFileContent, BaselineResult, UrlNormalizer } from './baseline-types';

@injectable()
export class BaselineGenerator {
    public generateBaseline(axeResultsList: AxeResultsList, urlNormalizer?: UrlNormalizer): BaselineFileContent {
        const combinedViolations = axeResultsList.values();
        const unsortedBaselineResults = combinedViolations.map((result) => this.combinedViolationToBaselineResult(result, urlNormalizer));
        const sortedBaselineResults = this.sortBaselineResults(unsortedBaselineResults);

        const baselineContent: BaselineFileContent = {
            metadata: { fileFormatVersion: '1' },
            results: sortedBaselineResults,
        };

        return baselineContent;
    }

    private combinedViolationToBaselineResult(result: AxeResult, urlNormalizer?: UrlNormalizer): BaselineResult {
        const node = result.junctionNode;
        if (node == null) {
            throw new Error('Invalid input result; does not contain a junctionNode');
        }

        const cssSelector = node.selectors.find((s) => s.type === 'css')?.selector;
        const xpathSelector = node.selectors.find((s) => s.type === 'xpath')?.selector;
        const urls = result.urls.map(urlNormalizer ?? identity).sort();

        if (cssSelector == null) {
            throw new Error('Invalid input result; does not contain a css selector');
        }

        // The order in which properties are added to the result matters; it affects the serialized baseline file,
        // which must be byte-for-byte consistent over time.
        const baselineResult: BaselineResult = {
            cssSelector,
            htmlSnippet: node.html,
            rule: result.id,
            urls,
        };

        // It's important to do this rather than just allowing "xpathSelector: undefined" in the output object.
        // This is because, again, to ensure that we serialize consistently.
        if (xpathSelector != null) {
            baselineResult.xpathSelector = xpathSelector;
        }

        return baselineResult;
    }

    private sortBaselineResults(results: BaselineResult[]): BaselineResult[] {
        return sortBy(results, ['rule', 'cssSelector', 'xpathSelector', 'htmlSnippet']);
    }
}
