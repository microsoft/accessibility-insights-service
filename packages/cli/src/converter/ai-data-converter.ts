// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { inject, injectable } from 'inversify';
import { CombinedReportDataConverter, ScanResultData, AxeCoreResults } from 'axe-result-converter';
import { CombinedReportParameters } from 'accessibility-insights-report';

@injectable()
export class AICombinedReportDataConverter {
    constructor(@inject(CombinedReportDataConverter) private readonly combinedReportDataConverter: CombinedReportDataConverter) {}

    public convertCrawlingResults(combinedAxeResults: AxeCoreResults, scanResultData: ScanResultData): CombinedReportParameters {
        return this.combinedReportDataConverter.convert(combinedAxeResults, scanResultData);
    }
}
