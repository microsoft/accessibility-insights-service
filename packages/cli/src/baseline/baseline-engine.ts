// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { UrlInfo } from 'accessibility-insights-report';
import { FingerprintGenerator } from '../converter/fingerprint-generator';
import { AxeCoreResults, AxeResult } from '../converter/axe-result-types';
import { BaselineEvaluation, BaselineOptions, BaselineResult, CountsByRule } from './baseline-types';
import { BaselineGenerator } from './baseline-generator';

interface UrlComparison {
    fixedCount: number;
    intersectingCount: number;
    newUrls: Set<string>;
    urlInfos: UrlInfo[];
}

const newComesFirst = -1;
const oldComesFirst = 1;

@injectable()
export class BaselineEngine {
    constructor(
        @inject(BaselineGenerator) private readonly baselineGenerator: BaselineGenerator,
        @inject(FingerprintGenerator) private readonly fingerprintGenerator: FingerprintGenerator,
    ) {}

    public updateResultsInPlace(axeResults: AxeCoreResults, baselineOptions: BaselineOptions): BaselineEvaluation {
        const oldBaselineResults: BaselineResult[] = baselineOptions.baselineContent?.results ?? [];
        const newBaseline = this.baselineGenerator.generateBaseline(axeResults.violations, baselineOptions.urlNormalizer);
        const newBaselineResults: BaselineResult[] = newBaseline.results;

        const evaluation: BaselineEvaluation = {
            suggestedBaselineUpdate: baselineOptions.baselineContent ? null : newBaseline,
            fixedViolationsByRule: {},
            newViolationsByRule: {},
            totalFixedViolations: 0,
            totalNewViolations: 0,
        };

        if (baselineOptions.baselineContent) {
            evaluation.totalBaselineViolations = 0;
        }

        // Take advantage of the fact that baseline results are always sorted
        let oldResultIndex = 0;
        let newResultIndex = 0;

        while (oldResultIndex < oldBaselineResults.length || newResultIndex < newBaselineResults.length) {
            const oldBaselineResult = oldResultIndex < oldBaselineResults.length ? oldBaselineResults[oldResultIndex] : null;
            const newBaselineResult = newResultIndex < newBaselineResults.length ? newBaselineResults[newResultIndex] : null;

            const resultDetailComparison = this.compareResultDetails(oldBaselineResult, newBaselineResult);

            if (resultDetailComparison < 0) {
                // exists in oldBaselineResults but not newBaselineResults
                this.addFixedViolationsToEvaluation(oldBaselineResult, evaluation);
                oldResultIndex++;
            } else if (resultDetailComparison > 0) {
                // exists in newBaselineResults but not oldBaselineResults
                this.addNewViolationsToEvaluation(newBaselineResult, evaluation);
                this.updateAxeResults(axeResults, newBaselineResult);
                newResultIndex++;
            } else {
                // exists in both oldBaselineResults and newBaselineResults, so compare urls
                const urlComparison: UrlComparison = this.getUrlComparison(oldBaselineResult.urls, newBaselineResult.urls);
                if (urlComparison.fixedCount) {
                    this.updateCountsByRule(evaluation.fixedViolationsByRule, oldBaselineResult.rule, urlComparison.fixedCount);
                    evaluation.totalFixedViolations += urlComparison.fixedCount;
                }
                if (urlComparison.newUrls.size) {
                    this.updateCountsByRule(evaluation.newViolationsByRule, oldBaselineResult.rule, urlComparison.newUrls.size);
                    evaluation.totalNewViolations += urlComparison.newUrls.size;
                }
                evaluation.totalBaselineViolations += oldBaselineResult.urls.length;

                this.updateAxeResults(axeResults, newBaselineResult, urlComparison.newUrls);

                oldResultIndex++;
                newResultIndex++;
            }
        }

        if (
            evaluation.totalFixedViolations ||
            evaluation.totalNewViolations ||
            baselineOptions.baselineContent?.metadata?.fileFormatVersion !== newBaseline.metadata.fileFormatVersion
        ) {
            evaluation.suggestedBaselineUpdate = newBaseline;
        }

        return evaluation;
    }

    private updateAxeResults(axeResults: AxeCoreResults, baselineResult: BaselineResult, newUrls?: Set<string>): void {
        const fingerprint = this.fingerprintGenerator.getFingerprint({
            rule: baselineResult.rule,
            snippet: baselineResult.htmlSnippet,
            cssSelector: baselineResult.cssSelector,
            xpathSelector: baselineResult.xpathSelector,
        });

        const resultToUpdate: AxeResult = axeResults.violations.get(fingerprint);
        resultToUpdate.urlInfos = this.buildUrlInfos(baselineResult, newUrls);
    }

    private buildUrlInfos(newBaselineResult: BaselineResult, newUrls?: Set<string>): UrlInfo[] {
        const urlInfos: UrlInfo[] = [];
        newBaselineResult.urls.map((url) => urlInfos.push(this.buildUrlInfo(url, newUrls)));

        return urlInfos;
    }

    private buildUrlInfo(url: string, newUrls?: Set<string>): UrlInfo {
        const isNew: boolean = !newUrls || newUrls.has(url);

        return {
            url,
            baselineStatus: isNew ? 'new' : 'existing',
        };
    }

    private addFixedViolationsToEvaluation(fixedViolation: BaselineResult, evaluation: BaselineEvaluation): void {
        this.updateCountsByRule(evaluation.fixedViolationsByRule, fixedViolation.rule, fixedViolation.urls.length);
        evaluation.totalFixedViolations += fixedViolation.urls.length;
        evaluation.totalBaselineViolations += fixedViolation.urls.length;
    }

    private addNewViolationsToEvaluation(newViolation: BaselineResult, evaluation: BaselineEvaluation): void {
        this.updateCountsByRule(evaluation.newViolationsByRule, newViolation.rule, newViolation.urls.length);
        evaluation.totalNewViolations += newViolation.urls.length;
    }

    private compareResultDetails(oldResult: BaselineResult | null, newResult: BaselineResult | null): number {
        if (oldResult && newResult) {
            // Compare the results in the order that they're sorted (rule, cssSelector, xPathSelector, htmlSnippet))
            return (
                this.safelyCompareStrings(oldResult.rule, newResult.rule) ||
                this.safelyCompareStrings(oldResult.cssSelector, newResult.cssSelector) ||
                this.safelyCompareStrings(oldResult.xpathSelector, newResult.xpathSelector) ||
                this.safelyCompareStrings(oldResult.htmlSnippet, newResult.htmlSnippet)
            );
        }

        return oldResult ? newComesFirst : oldComesFirst;
    }

    private safelyCompareStrings(oldString: string | undefined, newString: string | undefined): number {
        if (oldString && newString) {
            return oldString.localeCompare(newString);
        }

        if (!oldString && !newString) {
            return 0;
        }

        return oldString ? newComesFirst : oldComesFirst;
    }

    private updateCountsByRule(countsByRule: CountsByRule, rule: string, count: number): void {
        const oldCount = countsByRule[rule] || 0;

        countsByRule[rule] = oldCount + count;
    }

    private getUrlComparison(oldUrls: string[], newUrls: string[]): UrlComparison {
        const urlComparison: UrlComparison = {
            fixedCount: 0,
            intersectingCount: 0,
            newUrls: new Set(),
            urlInfos: [],
        };

        const oldSet: Set<string> = new Set(oldUrls);

        newUrls.map((url) => {
            if (oldSet.has(url)) {
                urlComparison.intersectingCount++;
                urlComparison.urlInfos.push({
                    url,
                    baselineStatus: 'existing',
                });
            } else {
                urlComparison.newUrls.add(url);
                urlComparison.urlInfos.push({
                    url,
                    baselineStatus: 'new',
                });
            }
        });

        urlComparison.fixedCount = oldSet.size - urlComparison.intersectingCount;

        return urlComparison;
    }
}
