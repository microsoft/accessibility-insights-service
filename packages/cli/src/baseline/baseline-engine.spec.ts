// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'reflect-metadata';

import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';
import { FingerprintGenerator, FingerprintParameters } from '../converter/fingerprint-generator';
import { AxeCoreResults, AxeResult, AxeResultsList } from '../converter/axe-result-types';
import { BaselineFileContent, BaselineOptions, UrlNormalizer } from './baseline-types';
import { BaselineEngine } from './baseline-engine';
import { BaselineGenerator } from './baseline-generator';

describe(BaselineEngine, () => {
    let baselineGeneratorMock: IMock<BaselineGenerator>;
    let fingerprintGeneratorMock: IMock<FingerprintGenerator>;
    let inputViolations: AxeResultsList;
    let inputResults: AxeCoreResults;
    let inputUrlNormalizer: UrlNormalizer;
    let inputOptions: BaselineOptions;
    let baselineContentWithNoViolations: BaselineFileContent;
    let baselineContentWithSingleViolation: BaselineFileContent;
    let baselineContentWithRuleViolationsBeforeAndAfterSingleViolation: BaselineFileContent;
    let baselineContentWithMultipleRuleViolations: BaselineFileContent;
    let baselineContentWithMutipleComplexChanges: BaselineFileContent;
    let testSubject: BaselineEngine;

    beforeEach(() => {
        baselineGeneratorMock = Mock.ofType<BaselineGenerator>(null, MockBehavior.Strict);
        fingerprintGeneratorMock = Mock.ofType<FingerprintGenerator>(null, MockBehavior.Strict);

        baselineContentWithNoViolations = {
            metadata: { fileFormatVersion: '1' },
            results: [],
        };
        baselineContentWithSingleViolation = {
            metadata: { fileFormatVersion: '1' },
            results: [
                {
                    cssSelector: '#some-id',
                    htmlSnippet: '<div id="some-id" />',
                    rule: 'rule-1',
                    urls: ['url-1'],
                },
            ],
        };
        baselineContentWithMultipleRuleViolations = {
            metadata: { fileFormatVersion: '1' },
            results: [
                {
                    cssSelector: '#some-id',
                    htmlSnippet: '<div id="some-id" />',
                    rule: 'rule-1',
                    urls: ['url-1', 'url-2'], // new URL
                },
                {
                    // new rule
                    cssSelector: '#another-id',
                    htmlSnippet: '<div id-"another-id" />',
                    rule: 'rule-2',
                    urls: ['url-2'],
                },
            ],
        };
        baselineContentWithMutipleComplexChanges = {
            metadata: { fileFormatVersion: '1' },
            results: [
                {
                    cssSelector: '#some-id',
                    htmlSnippet: '<div id="some-id" />',
                    rule: 'rule-1',
                    urls: ['url-1', 'url-2', 'url-3'], // new URL
                },
                {
                    // this rule replaces rule-2
                    cssSelector: '#yet-another-id',
                    htmlSnippet: '<div id-"yet-another-id" />',
                    rule: 'rule-3',
                    urls: ['url-3'],
                },
            ],
        };
        baselineContentWithRuleViolationsBeforeAndAfterSingleViolation = {
            metadata: { fileFormatVersion: '1' },
            results: [
                {
                    cssSelector: '#some-id-0',
                    htmlSnippet: '<div id="some-id-0" />',
                    rule: 'rule-0',
                    urls: ['url-0'],
                },
                {
                    cssSelector: '#some-id',
                    htmlSnippet: '<div id="some-id" />',
                    rule: 'rule-1',
                    urls: ['url-1'],
                },
                {
                    cssSelector: '#some-id-2',
                    htmlSnippet: '<div id="some-id-2" />',
                    rule: 'rule-2',
                    urls: ['url-2'],
                },
            ],
        };

        inputUrlNormalizer = (x) => x;
        inputOptions = {
            baselineContent: null,
            urlNormalizer: inputUrlNormalizer,
        };
        inputViolations = new AxeResultsList();
        inputResults = {
            violations: inputViolations,
        } as AxeCoreResults;

        testSubject = new BaselineEngine(baselineGeneratorMock.object, fingerprintGeneratorMock.object);
    });

    afterEach(() => {
        baselineGeneratorMock.verifyAll();
        fingerprintGeneratorMock.verifyAll();
    });

    const setupFingerprintMock = (): void => {
        fingerprintGeneratorMock
            .setup((m) => m.getFingerprint(It.isAny()))
            .returns((args: FingerprintParameters) => {
                const xpathPortion: string = args.xpathSelector ? `|${args.xpathSelector}` : '';

                return `${args.rule}|${args.snippet}|${args.cssSelector}${xpathPortion}`;
            })
            .verifiable(Times.atLeastOnce());
    };

    const setupCurrentScanContents = (baseline: BaselineFileContent): void => {
        for (const result of baseline.results) {
            const fingerprint = `${result.rule}|${result.htmlSnippet}|${result.cssSelector}`;
            inputResults.violations.add(fingerprint, {
                urls: result.urls,
                urlInfos: [],
                fingerprint,
            } as AxeResult);
        }

        baselineGeneratorMock.setup((m) => m.generateBaseline(inputResults.violations, inputUrlNormalizer)).returns(() => baseline);
    };

    const setupBaselineScanContents = (baseline: BaselineFileContent): void => {
        inputOptions.baselineContent = baseline;
    };

    describe('updateResultsInPlace', () => {
        it('propagates errors from the generator', () => {
            const generatorError = new Error('from BaselineGenerator');
            baselineGeneratorMock.setup((m) => m.generateBaseline(inputViolations, inputUrlNormalizer)).throws(generatorError);

            expect(() => testSubject.updateResultsInPlace(inputResults, inputOptions)).toThrowError(generatorError);
        });

        it('reports a different output if the file format version changes', () => {
            const baselineContentWithDifferentFileFormatVersion: BaselineFileContent = {
                metadata: { fileFormatVersion: '1.1' },
                results: [],
            } as unknown as BaselineFileContent;
            setupCurrentScanContents(baselineContentWithDifferentFileFormatVersion);
            setupBaselineScanContents(baselineContentWithNoViolations);

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithDifferentFileFormatVersion);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('Scan with no violations when no baseline content exists', () => {
            setupCurrentScanContents(baselineContentWithNoViolations);
            setupBaselineScanContents(null);

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithNoViolations);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('Scan with violations when no baseline content exists', () => {
            setupCurrentScanContents(baselineContentWithMultipleRuleViolations);
            setupBaselineScanContents(null);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithMultipleRuleViolations);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('current scan has more errors than baseline content', () => {
            setupCurrentScanContents(baselineContentWithMultipleRuleViolations);
            setupBaselineScanContents(baselineContentWithSingleViolation);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithMultipleRuleViolations);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('current scan content is identical to baseline content', () => {
            setupCurrentScanContents(baselineContentWithMultipleRuleViolations);
            setupBaselineScanContents(baselineContentWithMultipleRuleViolations);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBeNull();
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('current scan content has fewer errors than baseline content', () => {
            setupCurrentScanContents(baselineContentWithSingleViolation);
            setupBaselineScanContents(baselineContentWithMultipleRuleViolations);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithSingleViolation);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('current scan content has complex differences compared to baseline content', () => {
            setupCurrentScanContents(baselineContentWithMutipleComplexChanges);
            setupBaselineScanContents(baselineContentWithMultipleRuleViolations);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithMutipleComplexChanges);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });

        it('current scan content has new failures before and after baseline content', () => {
            setupCurrentScanContents(baselineContentWithRuleViolationsBeforeAndAfterSingleViolation);
            setupBaselineScanContents(baselineContentWithSingleViolation);
            setupFingerprintMock();

            const evaluation = testSubject.updateResultsInPlace(inputResults, inputOptions);

            expect(evaluation.suggestedBaselineUpdate).toBe(baselineContentWithRuleViolationsBeforeAndAfterSingleViolation);
            expect({ evaluation: evaluation, axeResults: inputResults }).toMatchSnapshot();
        });
    });
});
