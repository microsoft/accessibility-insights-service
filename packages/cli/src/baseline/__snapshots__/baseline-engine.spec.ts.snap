// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaselineEngine updateResultsInPlace Scan with no violations when no baseline content exists 1`] = `
{
  "axeResults": {
    "violations": {},
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {},
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [],
    },
    "totalFixedViolations": 0,
    "totalNewViolations": 0,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace Scan with violations when no baseline content exists 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-1",
          },
          {
            "baselineStatus": "new",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-1",
          "url-2",
        ],
      },
      "rule-2|<div id-"another-id" />|#another-id": {
        "fingerprint": "rule-2|<div id-"another-id" />|#another-id",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-2",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {
      "rule-1": 2,
      "rule-2": 1,
    },
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [
        {
          "cssSelector": "#some-id",
          "htmlSnippet": "<div id="some-id" />",
          "rule": "rule-1",
          "urls": [
            "url-1",
            "url-2",
          ],
        },
        {
          "cssSelector": "#another-id",
          "htmlSnippet": "<div id-"another-id" />",
          "rule": "rule-2",
          "urls": [
            "url-2",
          ],
        },
      ],
    },
    "totalFixedViolations": 0,
    "totalNewViolations": 3,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace current scan content has complex differences compared to baseline content 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-1",
          },
          {
            "baselineStatus": "existing",
            "url": "url-2",
          },
          {
            "baselineStatus": "new",
            "url": "url-3",
          },
        ],
        "urls": [
          "url-1",
          "url-2",
          "url-3",
        ],
      },
      "rule-3|<div id-"yet-another-id" />|#yet-another-id": {
        "fingerprint": "rule-3|<div id-"yet-another-id" />|#yet-another-id",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-3",
          },
        ],
        "urls": [
          "url-3",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {
      "rule-2": 1,
    },
    "newViolationsByRule": {
      "rule-1": 1,
      "rule-3": 1,
    },
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [
        {
          "cssSelector": "#some-id",
          "htmlSnippet": "<div id="some-id" />",
          "rule": "rule-1",
          "urls": [
            "url-1",
            "url-2",
            "url-3",
          ],
        },
        {
          "cssSelector": "#yet-another-id",
          "htmlSnippet": "<div id-"yet-another-id" />",
          "rule": "rule-3",
          "urls": [
            "url-3",
          ],
        },
      ],
    },
    "totalBaselineViolations": 3,
    "totalFixedViolations": 1,
    "totalNewViolations": 2,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace current scan content has fewer errors than baseline content 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-1",
          },
        ],
        "urls": [
          "url-1",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {
      "rule-1": 1,
      "rule-2": 1,
    },
    "newViolationsByRule": {},
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [
        {
          "cssSelector": "#some-id",
          "htmlSnippet": "<div id="some-id" />",
          "rule": "rule-1",
          "urls": [
            "url-1",
          ],
        },
      ],
    },
    "totalBaselineViolations": 3,
    "totalFixedViolations": 2,
    "totalNewViolations": 0,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace current scan content has new failures before and after baseline content 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-0|<div id="some-id-0" />|#some-id-0": {
        "fingerprint": "rule-0|<div id="some-id-0" />|#some-id-0",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-0",
          },
        ],
        "urls": [
          "url-0",
        ],
      },
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-1",
          },
        ],
        "urls": [
          "url-1",
        ],
      },
      "rule-2|<div id="some-id-2" />|#some-id-2": {
        "fingerprint": "rule-2|<div id="some-id-2" />|#some-id-2",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-2",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {
      "rule-0": 1,
      "rule-2": 1,
    },
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [
        {
          "cssSelector": "#some-id-0",
          "htmlSnippet": "<div id="some-id-0" />",
          "rule": "rule-0",
          "urls": [
            "url-0",
          ],
        },
        {
          "cssSelector": "#some-id",
          "htmlSnippet": "<div id="some-id" />",
          "rule": "rule-1",
          "urls": [
            "url-1",
          ],
        },
        {
          "cssSelector": "#some-id-2",
          "htmlSnippet": "<div id="some-id-2" />",
          "rule": "rule-2",
          "urls": [
            "url-2",
          ],
        },
      ],
    },
    "totalBaselineViolations": 1,
    "totalFixedViolations": 0,
    "totalNewViolations": 2,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace current scan content is identical to baseline content 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-1",
          },
          {
            "baselineStatus": "existing",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-1",
          "url-2",
        ],
      },
      "rule-2|<div id-"another-id" />|#another-id": {
        "fingerprint": "rule-2|<div id-"another-id" />|#another-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-2",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {},
    "suggestedBaselineUpdate": null,
    "totalBaselineViolations": 3,
    "totalFixedViolations": 0,
    "totalNewViolations": 0,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace current scan has more errors than baseline content 1`] = `
{
  "axeResults": {
    "violations": {
      "rule-1|<div id="some-id" />|#some-id": {
        "fingerprint": "rule-1|<div id="some-id" />|#some-id",
        "urlInfos": [
          {
            "baselineStatus": "existing",
            "url": "url-1",
          },
          {
            "baselineStatus": "new",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-1",
          "url-2",
        ],
      },
      "rule-2|<div id-"another-id" />|#another-id": {
        "fingerprint": "rule-2|<div id-"another-id" />|#another-id",
        "urlInfos": [
          {
            "baselineStatus": "new",
            "url": "url-2",
          },
        ],
        "urls": [
          "url-2",
        ],
      },
    },
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {
      "rule-1": 1,
      "rule-2": 1,
    },
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1",
      },
      "results": [
        {
          "cssSelector": "#some-id",
          "htmlSnippet": "<div id="some-id" />",
          "rule": "rule-1",
          "urls": [
            "url-1",
            "url-2",
          ],
        },
        {
          "cssSelector": "#another-id",
          "htmlSnippet": "<div id-"another-id" />",
          "rule": "rule-2",
          "urls": [
            "url-2",
          ],
        },
      ],
    },
    "totalBaselineViolations": 1,
    "totalFixedViolations": 0,
    "totalNewViolations": 2,
  },
}
`;

exports[`BaselineEngine updateResultsInPlace reports a different output if the file format version changes 1`] = `
{
  "axeResults": {
    "violations": {},
  },
  "evaluation": {
    "fixedViolationsByRule": {},
    "newViolationsByRule": {},
    "suggestedBaselineUpdate": {
      "metadata": {
        "fileFormatVersion": "1.1",
      },
      "results": [],
    },
    "totalBaselineViolations": 0,
    "totalFixedViolations": 0,
    "totalNewViolations": 0,
  },
}
`;
